[scroll-animations] Have ViewTimeline hold a Styleable subject

nmoucht opened the pull request:
#### f06878c33bc9f63117a77f6c67a5b5f4566f94f3
<pre>
[scroll-animations] Have ViewTimeline hold a Styleable subject
<a href="https://bugs.webkit.org/show_bug.cgi?id=286723">https://bugs.webkit.org/show_bug.cgi?id=286723</a>
<a href="https://rdar.apple.com/143860366">rdar://143860366</a>

Reviewed by Antoine Quint.

To handle ViewTimelines with associated pseudo elements, have ViewTimeline hold a
styleable subject.

* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::AnimationTimelinesController::registerNamedViewTimeline):
* Source/WebCore/animation/AnimationTimelinesController.h:
* Source/WebCore/animation/CSSAnimation.cpp:
(WebCore::CSSAnimation::syncStyleOriginatedTimeline):
* Source/WebCore/animation/ScrollTimeline.h:
* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::ViewTimeline::subject const):
(WebCore::ViewTimeline::setSubject):
(WebCore::ViewTimeline::controller const):
(WebCore::ViewTimeline::cacheCurrentTime):
(WebCore::ViewTimeline::documentWillUpdateAnimationsAndSendEvents):
(WebCore::ViewTimeline::sourceScrollerRenderer const):
* Source/WebCore/animation/ViewTimeline.h:

Canonical link: <a href="https://commits.webkit.org/289801@main">https://commits.webkit.org/289801@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/803b451d08939674f852ff77a4d5fd155fa9f266

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87963 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7479 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42391 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92866 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38713 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90014 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7860 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15655 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67900 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25639 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90965 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6021 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79615 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48269 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5797 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/34023 "Found 1 new test failure: accessibility/noscript-ignored.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37820 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76196 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34904 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94728 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15131 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11128 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76753 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15386 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75471 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75991 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20382 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18825 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/8135 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13737 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15149 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20450 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14891 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18336 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16673 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/view-timeline-pseudo-on-scroller.html:
```diff
@@ -0,0 +1,62 @@
+  width: 100px;
+  height: 50px;
+  background: red;
+  animation: bg linear reverse;
```
I don't think `reverse` here is material to the test.

nmoucht replied:
Removed.

graouts commented on Source/WebCore/animation/CSSAnimation.cpp:
```diff
@@ -168,7 +168,10 @@ void CSSAnimation::syncStyleOriginatedTimeline()
+            if (auto owningElement = this->owningElement())
+                viewTimeline->setSubject(*owningElement);
+            else
+                viewTimeline->setSubject(nullptr);
```
We have an `ASSERT(owningElement())` at the start of this method, so presumably we shouldn't enter the `else` branch here. We should also fix the similar code for scroll timelines.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -155,31 +155,46 @@ ExceptionOr<ViewTimeline::SpecifiedViewTimelineInsets> ViewTimeline::validateSpe
+    if (!subject)
+        return nullptr;
+
+    return &subject->element;
```
```suggestion
    if (auto subject = m_subject.styleable())
        return &subject->element;
    return nullptr;
```

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -155,31 +155,46 @@ ExceptionOr<ViewTimeline::SpecifiedViewTimelineInsets> ViewTimeline::validateSpe
+    return &subject->element;
+}
+
+
```
Extra empty line.

nmoucht replied:
Removed.

graouts commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -240,10 +256,10 @@ void ViewTimeline::cacheCurrentTime()
 
         if (m_specifiedInsets) {
-            RefPtr subject { m_subject.get() };
+            RefPtr subjectPtr { &subject->element };
```
Not sure there's much value in renaming this parameter.

nmoucht replied:
Its necessary as subject is used to hold the styleable now.

graouts replied:
Perhaps name it in a more descriptive fashion then. It's not so much that it's a pointer that's relevant but rather than it's the subject's element. So `subjectElement`?

graouts reviewed (approved):
We need to export the new test to WPT.

nmoucht committed:
[scroll-animations] Have ViewTimeline hold a Styleable subject
https://bugs.webkit.org/show_bug.cgi?id=286723
rdar://143860366

Reviewed by Antoine Quint.

To handle ViewTimelines with associated pseudo elements, have ViewTimeline hold a
styleable subject.

* Source/WebCore/animation/AnimationTimelinesController.cpp:
(WebCore::AnimationTimelinesController::registerNamedViewTimeline):
* Source/WebCore/animation/AnimationTimelinesController.h:
* Source/WebCore/animation/CSSAnimation.cpp:
(WebCore::CSSAnimation::syncStyleOriginatedTimeline):
* Source/WebCore/animation/ScrollTimeline.h:
* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::ViewTimeline::subject const):
(WebCore::ViewTimeline::setSubject):
(WebCore::ViewTimeline::controller const):
(WebCore::ViewTimeline::cacheCurrentTime):
(WebCore::ViewTimeline::documentWillUpdateAnimationsAndSendEvents):
(WebCore::ViewTimeline::sourceScrollerRenderer const):
* Source/WebCore/animation/ViewTimeline.h:

Canonical link: https://commits.webkit.org/289801@main
